### PR TITLE
Switch to materialization job type (from engine) and add strategy

### PR DESCRIPTION
**Pull request by shangyian**
#### Materialization Job Type

This adds the concept of a materialization `job type`, which moves away from the original materialization `engine` concept. The reason behind this is that `engine` doesn't accurately describe each materialization job, and some materialization jobs can use more than one engine (for example, the `DruidCube` job uses two). 

Screenshot 2023-11-21 at 8 23 43 AM

#### Materialization Strategy

This also adds a materialization `strategy` field, which describes the type of materialization being done. This allows the user to explicitly choose a strategy, rather than us implicitly picking one based on the configuration. It also makes for clearer error checking for different strategies.

### Test Plan

Locally + unit test updates

- [ ] PR has an associated issue: #
- [ ] `make check` passes
- [ ] `make test` shows 100% unit test coverage

### Deployment Plan

Will require schema changes to the `materialization` table

### CODE REVIEW

File: datajunction-clients/python/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@ lint:
 	make check
 
 test:
-	pdm run pytest --cov=datajunction --cov-report=html -vv tests/ --doctest-modules datajunction --without-integration --without-slow-integration ${PYTEST_ARGS}
```
Comment (shangyian) [inline datajunction-clients/python/Makefile:8]: This runs the Python client tests in parallel

**Review by samredai — APPROVED** [review]
This is awesome, thanks @shangyian! I think we've really gotten to the right amount of user control while still making it feel declarative. Pick a job type then pick a strategy is a great flow and leaves a lot of room to extend in the future with other patterns.

File: datajunction-clients/python/Makefile (hunk 1 above)
Comment (agorajek) [reply datajunction-clients/python/Makefile:8]: Did you mean it will run the tests in parallel after your code change?

File: datajunction-server/datajunction_server/models/materialization.py [hunk 2]
```diff
@@ -41,6 +44,34 @@
...
+
+    # Typically used for smaller transforms tables, where for each run we
+    # fully overwrite the table.
+    # Availability state: single table
+    FULL = "full"
+
+    # Full materialization into a new snapshot for each scheduled materialization run.
+    # DJ will manage the location of the snapshot tables and return the appropriate snapshot
+    # when requested
+    # Availability state: multiple tables, one per snapshot. The snapshot key should be based on
```
Comment (agorajek) [inline datajunction-server/datajunction_server/models/materialization.py:60]: For dimension data tracking (and for other purposes) we could have a single table with a SNAPSHOT_PARTITION here. I have done both of these options (e.g. in the TableSnapshot job type) and you can also see that described in https://docs.google.com/document/d/1B5wu-1pgjvEqtG8Q9-4Nfi7DH1dJclOOXBpBvYV5pbk/edit#heading=h.qamncy4l751q under `_day_d` suffix pattern. 

With a single (partitioned) table there could always be a problem of a schema mismatch if the schema changes, but I solved it by creating a new snapshot table in such case.

Long story short... now you know, but feel free to keep it simple for now.

File: datajunction-clients/python/Makefile (hunk 1 above)
Comment (shangyian) [reply datajunction-clients/python/Makefile:8]: Yep! Although this only works locally (on github we just have 1 core, so the parallelism will just be 1)

File: datajunction-clients/python/datajunction/models.py [hunk 3]
```diff
@@ -14,12 +14,33 @@ class Engine(BaseModel):
...
+
+class MaterializationStrategy(str, enum.Enum):
+    """
+    Materialization strategies
+    """
+
+    FULL = "full"
+    SNAPSHOT = "snapshot"
+    INCREMENTAL_TIME = "incremental_time"
+    VIEW = "view"
```
[the remark this answers is not in the document]
Comment (agorajek) [reply datajunction-clients/python/datajunction/models.py:34]: I am sorry. I wrote it before I saw the docs in `[materialization.py](https://github.com/DataJunction/dj/pull/865/files/e293e0e009dac3a3c05c688fb20cb71adb7b3105#diff-e54aff7b15bfaf879d6b12d3cad7a285050df33dc65d976f2c034e3adb8c40da)`

File: datajunction-server/datajunction_server/models/materialization.py [hunk 4]
```diff
@@ -114,6 +145,11 @@ class GenericMaterializationConfigInput(BaseModel):
     # Spark config
     spark: Optional[SparkConf]
 
+    # The time window to lookback when overwriting materialized datasets
+    # This will only be used if a time partition was set on the node and
+    # the materialization strategy is INCREMENTAL_TIME
+    lookback_window: Optional[str]
```
[the remark this answers is not in the document]
Comment (shangyian) [reply datajunction-server/datajunction_server/models/materialization.py:159]: Good point, I was actually thinking that this would be something like `1 DAY`, which is parseable via Spark SQL's time string parser. I'll set that as the default.

[the remark this answers is not in the document]
Comment (agorajek) [reply datajunction-server/datajunction_server/models/materialization.py:159]: I just looked at your example form image and realized one thing about this. I think this makes most sense for the incremental_time, but for all the rest it should probably default to all-data in the source table. I think we should walk through these examples together... or I'll write them up after my next meeting.

[the remark this answers is not in the document]
Comment (agorajek) [reply datajunction-server/datajunction_server/models/materialization.py:159]: Ideally the form elements would react to the values of others but for now maybe we could just do some defaults and  describe the assumptions somewhere. Maybe for now we should let the users do whatever they want but eventually we should probably have more guide rails. Example:

Strategy:     
- INCREMENTAL_TIME:
   - LookBackWindow:   Any* (default: 1)                      
   - Schedule:    Any
- FULL:
   - LookBackWindow:   Any* (default: all)      
   - Schedule:    Disabled (No Backfill Needed)
- SNAPSHOT:
   - LookBackWindow:   Any* (default: all)      
   - Schedule:    Any (because it goes into a fixes snapshot space)
- VIEW:
   - LookBackWindow:   Any
   - Schedule:    Any (but probably default to @daily)

Look-back Window of Any* means that user can specify any INT value, but only if a temporal partition exists.

[the remark this answers is not in the document]
Comment (shangyian) [reply datajunction-server/datajunction_server/models/materialization.py:159]: Agreed -- the UI should be changed so that if someone selects "incremental time", it'll display a lookback window form field, but if they select any other thing, it won't allow the option for lookback window. 

I also just added some parsing to handle different lookback window values, which are actually just any Spark SQL compatible `INTERVAL` values (someone could specify it to be `1 day` or `5 hours 30 seconds 1 milliseconds 1 microseconds`, for example, and Spark will handle the parsing).
